lnwallet/chanfunding/psbt_assembler.go

```diff
@@ -0,0 +1,372 @@
+	// outputs haven't changed to what was previously verified. Only witness
+	// data should be added after the verification process.
+	PendingPsbt *psbt.Packet
+
+	// PsbtReady is an error channel the funding manager will listen for
+	// a signal about the PSBT being ready to continue the funding flow. In
+	// the normal, happy flow, this channel is only ever closed. If a
+	// non-nil error is sent through the channel, the funding flow will be
+	// canceled.
+	PsbtReady chan error
+
+	// signalPsbtReady is a Once guard to make sure the PsbtReady channel is
+	// only closed exactly once.
+	signalPsbtReady sync.Once
+}
+
+// BindKeys sets both the remote and local node's keys that will be used for the
+// channel funding multisig output.
+func (i *PsbtIntent) BindKeys(localKey *keychain.KeyDescriptor,
+	remoteKey *btcec.PublicKey) {
+
+	i.localKey = localKey
+	i.remoteKey = remoteKey
+	i.State = PsbtOutputKnown
+}
+
+// FundingParams returns the parameters that are necessary to start funding the
+// channel output this intent was created for. It returns the bech32 encoded
+// P2WSH funding address, the exact funding amount and a raw PSBT that contains
+// exactly one output that encodes the previous two parameters.
+func (i *PsbtIntent) FundingParams(netParams *chaincfg.Params) (string, int64,
+	[]byte, error) {
+
+	// The funding output needs to be known already at this point, which
+	// means we need to have the local and remote multisig keys bound
+	// already.
+	witnessScript, out, err := i.FundingOutput()
+	if err != nil {
+		return "", 0, nil, fmt.Errorf("unable to create funding "+
+			"output: %v", err)
+	}
+	witnessScriptHash := sha256.Sum256(witnessScript)
+
+	// Encode the address in the human readable bech32 format.
+	addr, err := btcutil.NewAddressWitnessScriptHash(
+		witnessScriptHash[:], netParams,
+	)
+	if err != nil {
+		return "", 0, nil, fmt.Errorf("unable to encode address: %v",
+			err)
+	}
+
+	// We'll also encode the address/amount in a machine readable raw PSBT
+	// format. If the user supplied a base PSBT, we'll add the output to
+	// that one, otherwise we'll create a new one.
+	packet := i.BasePsbt
+	if packet == nil {
+		packet, err = psbt.New(nil, nil, 2, 0, nil)
+		if err != nil {
+			return "", 0, nil, fmt.Errorf("unable to create PSBT: "+
+				"%v", err)
+		}
+	}
+	packet.UnsignedTx.TxOut = append(packet.UnsignedTx.TxOut, out)
+	packet.Outputs = append(packet.Outputs, psbt.POutput{})
+
+	// Serialize the new or updated PSBT into the raw binary format.
+	var buf bytes.Buffer
+	err = packet.Serialize(&buf)
+	if err != nil {
+		return "", 0, nil, fmt.Errorf("unable to serialize PSBT: %v",
+			err)
+	}
+
+	return addr.EncodeAddress(), out.Value, buf.Bytes(), nil
+}
+
+// Verify makes sure the PSBT that is given to the intent has an output that
+// sends to the channel funding multisig address with the correct amount. A
+// simple check that there is a large enough input present is also performed.
```

Comment:
> there is a large enough input present is also performed

I think this instead should be: "that at least a single input has been specified is performed". Assuming we don't have access to the UTXO state in this context, we can't check the input UTXO size.

Reply:
Right, thanks. I wrote the comment before I realized we don't have the UTXO information here.